src/nomad_simulations/properties/band_structure.py

```diff
@@ -0,0 +1,377 @@
+        unit='joule',
+        description="""
+        Highest occupied electronic eigenvalue. Together with `lowest_unoccupied`, it defines the
+        electronic band gap.
+        """,
+    )
+
+    lowest_unoccupied = Quantity(
+        type=np.float64,
+        unit='joule',
+        description="""
+        Lowest unoccupied electronic eigenvalue. Together with `highest_occupied`, it defines the
+        electronic band gap.
+        """,
+    )
+
+    value_contributions = SubSection(
+        sub_section=BaseElectronicEigenvalues.m_def,
+        repeats=True,
+        description="""
+        Contributions to the electronic eigenvalues. Example, in the case of a DFT+GW calculation, the GW eigenvalues
+        are stored under `value`, and each contribution is identified by `label`:
+            - `'KS'`: Kohn-Sham contribution. This is also stored in the DFT entry under `ElectronicEigenvalues.value`.
+            - `'KSxc'`: Diagonal matrix elements of the expectation value of the Kohn-Sahm exchange-correlation potential.
+            - `'SigX'`: Diagonal matrix elements of the exchange self-energy. This is also stored in the GW entry under `ElectronicSelfEnergy.value`.
+            - `'SigC'`: Diagonal matrix elements of the correlation self-energy. This is also stored in the GW entry under `ElectronicSelfEnergy.value`.
+            - `'Zk'`: Quasiparticle renormalization factors contribution. This is also stored in the GW entry under `QuasiparticleWeights.value`.
+        """,
+    )
+
+    reciprocal_cell = Quantity(
+        type=KSpace.reciprocal_lattice_vectors,
+        description="""
+        Reference to the reciprocal lattice vectors stored under `KSpace`. This reference is useful when resolving the Brillouin zone
+        for the front-end visualization.
+        """,
+    )
+
+    def __init__(
+        self, m_def: Section = None, m_context: Context = None, **kwargs
+    ) -> None:
+        super().__init__(m_def, m_context, **kwargs)
+        self.name = self.m_def.name
+
+    def validate_occupation(self, logger: BoundLogger) -> bool:
+        """
+        Validate `occupation` by checking if they exist and have the same shape as `value`.
+
+        Args:
+            logger (BoundLogger): The logger to log messages.
+
+        Returns:
+            (bool): True if the shape of `occupation` is the same as `value`, False otherwise.
+        """
+        if self.occupation is None or len(self.occupation) == 0:
+            logger.warning('Cannot find `occupation` defined.')
+            return False
+        if self.value is not None and self.value.shape != self.occupation.shape:
+            logger.warning(
+                'The shape of `value` and `occupation` are different. They should have the same shape.'
+            )
+            return False
+        return True
+
+    def order_eigenvalues(
+        self, logger: BoundLogger
+    ) -> Tuple[Optional[pint.Quantity], Optional[np.ndarray]]:
```

Comment:
Problem comes with the units of the returned `sorted_value`.

Reply:
But, empty lists can also have units...

Reply:
Yeah, but knowing the unit implies that the extraction works. I think None and Optional is much more clean. I also doubt that an empty pint list is not screwing things on the archive, meaning, that the quantity is set to an empty value instead of being None, which is the behavior we want

Reply:
> I also doubt that an empty pint list is not screwing things on the archive, meaning, that the quantity is set to an empty value instead of being None, which is the behavior we want

That would be good to know, actually! Let's check it out. Tbh, the archive should have a way of handling empy arrays.

Reply:
> I think None and Optional is much more clean.

They are definitely more verbose, which indeed can be more communicative.
I'd call simpler logic _cleaner_, though. If a type (e.g. `list`, `int`) can handle "failure" (e.g. `[]`, `0`) by itself, I wouldn't complicate matters.
This is, of course, very situation-dependent, as sometimes "failure" may present itself differently, e.g. `1 / 0`.

Honestly, Python's way of handling `None` is pretty naive. In a language like Haskell I'd make less of an issue, as the logic is better separated. There, `Nothing` is a constructor of the **monad** `Maybe`, which handles "failure" propogation in a more automated and modular fashion.